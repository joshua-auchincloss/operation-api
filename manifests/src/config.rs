use std::path::PathBuf;

use config::File;
use serde::de::DeserializeOwned;
use validator::Validate;

use crate::Error;

pub trait NewForConfig
where
    Self: Sized + DeserializeOwned + Validate, {
    const NAME: &'static str;
    fn new<'a, S: Into<&'a str>>(dir: Option<S>) -> crate::Result<Self> {
        let file_name = format!(
            "{}",
            PathBuf::from(dir.map(|s| s.into()).unwrap_or("./"))
                .join(Self::NAME)
                .display()
        );

        let this: Self = config::ConfigBuilder::<config::builder::DefaultState>::default()
            .add_source(File::with_name(&file_name).required(true))
            .add_source(config::Environment::default().prefix("OP"))
            .build()
            .map_err(Error::from_with_source_init(file_name.clone()))?
            .try_deserialize()
            .map_err(Error::from_with_source_init(file_name.clone()))?;

        this.validate()
            .map_err(Error::from_with_source_init(file_name.clone()))?;

        Ok(this)
    }
}
