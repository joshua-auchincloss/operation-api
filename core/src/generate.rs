pub mod matcher;
pub mod remote;
pub mod rust;

use std::{fmt::Debug, path::PathBuf};

use config::File;
use serde::Deserialize;

use crate::{
    Definitions, Operation, Struct,
    context::Context,
    generate::{remote::RemoteConfig, rust::RustGenerator},
    namespace::WithNsContext,
};

#[derive(thiserror::Error, Debug)]
pub enum GenerationError {
    #[error("config error: {0}")]
    Config(#[from] config::ConfigError),

    #[error("glob error: {0}")]
    Glob(#[from] glob::GlobError),
}

type Result<T, E = GenerationError> = std::result::Result<T, E>;

#[derive(Deserialize, PartialEq, Debug)]
#[serde(rename_all = "snake_case")]
pub enum Language {
    Rust,
}

#[derive(Deserialize, PartialEq, Debug)]
#[serde(rename_all = "snake_case")]
pub enum Target {
    Client,
    Server,
    Types,
}

pub trait ConfigExt: Debug + PartialEq + Clone {}

#[derive(Deserialize, PartialEq, Debug, Clone)]
#[serde(rename_all = "kebab-case")]
pub struct GenOpts<Ext: ConfigExt> {
    output_dir: PathBuf,
    opts: Ext,
}

#[derive(Deserialize, PartialEq, Debug, Clone)]
#[serde(rename_all = "kebab-case")]
pub struct RustConfig {}

impl ConfigExt for RustConfig {}

crate::default!(
    Vec<Target>: { targets = vec![Target::Types] },
);

#[derive(Deserialize, PartialEq, Debug)]
pub struct Source {
    #[serde(default)]
    remote: Vec<RemoteConfig>,
    #[serde(default)]
    include: Vec<String>,
    #[serde(default)]
    exclude: Vec<String>,
}

#[derive(Deserialize, PartialEq, Debug)]
#[serde(rename_all = "kebab-case")]
pub struct GenerationConfig {
    sources: Source,

    #[serde(default = "default_targets")]
    targets: Vec<Target>,

    #[serde(default)]
    languages: Vec<Language>,

    #[serde(default)]
    rust: Option<GenOpts<RustConfig>>,
}

impl GenerationConfig {
    pub fn new<'a, S: Into<&'a str>>(dir: Option<S>) -> Result<Self> {
        Ok(
            config::ConfigBuilder::<config::builder::DefaultState>::default()
                .add_source(
                    File::with_name(&format!(
                        "{}",
                        PathBuf::from(dir.map(|s| s.into()).unwrap_or("./"))
                            .join("op-gen")
                            .display()
                    ))
                    .required(true),
                )
                .add_source(config::Environment::default().prefix("OP"))
                .build()?
                .try_deserialize()?,
        )
    }
}

pub trait Generate<State, Ext: ConfigExt> {
    fn new_state<'ns>(
        &self,
        opts: &GenOpts<Ext>,
    ) -> State;

    fn gen_ctx<'ns>(
        &self,
        ctx: &'ns Context,
        opts: &GenOpts<Ext>,
    ) -> Result<()> {
        let mut state = self.new_state(opts);

        for ns in ctx.namespaces.values() {
            let mut ns_ctx = WithNsContext::new(ns, &mut state);

            for def in ns.defs.values() {
                self.gen_struct(&mut ns_ctx, def)?;
            }

            for op in ns.ops.values() {
                self.gen_operation(&mut ns_ctx, op)?;
            }
        }

        Ok(())
    }

    fn gen_operation<'ns>(
        &self,
        state: &mut WithNsContext<'ns, State>,
        def: &Operation,
    ) -> Result<()>;

    fn gen_struct<'ns>(
        &self,
        state: &mut WithNsContext<'ns, State>,
        def: &Struct,
    ) -> Result<()>;
}

impl GenerationConfig {
    fn sources(&self) -> Result<Vec<PathBuf>> {
        matcher::walk(&self.sources)
    }

    pub fn get_ctx(&self) -> crate::Result<Context> {
        let mut ctx = Context::new();

        for source in self.sources()? {
            ctx.load_from_source(source)?;
        }

        ctx.finish()?;

        Ok(ctx)
    }

    pub async fn generate_all(&self) -> crate::Result<()> {
        let ctx = self.get_ctx()?;

        let mut futs = vec![];
        if let Some(rust) = &self.rust {
            let generator = RustGenerator;
            futs.push(Box::pin(async move { generator.gen_ctx(&ctx, rust) }));
        }

        for fut in futs {
            let _ = fut.await?;
        }

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[tokio::test]
    async fn test_config_loader() {
        let conf = GenerationConfig::new(Some("../samples/config-a")).unwrap();

        assert_eq! {
            conf, GenerationConfig {
                targets: vec![Target::Client, Target::Server, Target::Types],
                languages: vec![Language::Rust],
                sources: Source {
                    remote: vec![
                        RemoteConfig {
                            url: "http://localhost:9009/dynamic.toml".into(),
                            headers: Default::default()
                        }
                    ],
                    include: vec![
                        "../samples/basic-op.toml".into(),
                        "../samples/basic-struct.toml".into(),
                        "../samples/test-struct*.toml".into()
                    ],
                    exclude: vec![
                        "*basic-op*".into()
                    ]
                },
                rust: Some(GenOpts{
                    output_dir: "../samples/gen-a".into(),
                    opts: RustConfig {  }
                })
            }
        }

        assert_eq! {
            conf.sources().unwrap(),
            vec![
                PathBuf::from("../samples/basic-struct.toml"),
                PathBuf::from("../samples/test-struct-readme.toml"),
                PathBuf::from("../samples/test-struct-text.toml"),
            ]
        }

        let ctx = conf.get_ctx().unwrap();

        insta::assert_debug_snapshot!(ctx.namespaces);

        conf.generate_all().await.unwrap();
    }
}
