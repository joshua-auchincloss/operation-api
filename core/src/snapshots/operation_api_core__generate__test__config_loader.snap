---
source: core/src/generate.rs
expression: ctx.namespaces
---
abc.corp.namespace:
  name: abc.corp.namespace
  version: 1
  fields: {}
  ops: {}
  defs:
    some_struct:
      name: some_struct
      namespace: abc.corp.namespace
      description: ~
      version: 2
      fields:
        a:
          name: ~
          namespace: ~
          description: ~
          version: ~
          type: i32
          optional: false
        b:
          name: ~
          namespace: ~
          description: ~
          version: ~
          type: f32
          optional: true
        c:
          name: ~
          namespace: ~
          description: ~
          version: ~
          type:
            compound_type:
              kind: sized_array
              size: 4
              type:
                compound_type:
                  kind: sized_array
                  size: 4
                  type: f32
          optional: false
        d:
          name: ~
          namespace: ~
          description: ~
          version: ~
          type:
            compound_type:
              kind: sized_array
              size: 4
              type:
                compound_type:
                  kind: sized_array
                  size: 4
                  type: f32
          optional: false
  enums:
    some_enum:
      name: some_enum
      namespace: abc.corp.namespace
      description: some enum description
      version: 1
      variants:
        A:
          name: A
          namespace: ~
          description: ~
          version: ~
          value: 1
        B:
          name: B
          namespace: ~
          description: Some b variant = 2
          version: ~
          value: 2
abc.corp.test:
  name: abc.corp.test
  version: 1
  fields: {}
  ops: {}
  defs:
    BasicStruct:
      name: BasicStruct
      namespace: abc.corp.test
      description: Some struct
      version: 1
      fields:
        a:
          name: a
          namespace: abc.corp.test
          description: field a
          version: ~
          type: i32
          optional: false
        b:
          name: b
          namespace: abc.corp.test
          description: ~
          version: ~
          type:
            compound_type:
              kind: option
              type: i32
          optional: false
        c:
          name: c
          namespace: abc.corp.test
          description: ~
          version: ~
          type:
            compound_type:
              kind: array
              type: f32
          optional: false
        d:
          name: d
          namespace: abc.corp.test
          description: ~
          version: ~
          type:
            compound_type:
              kind: sized_array
              size: 4
              type:
                compound_type:
                  kind: sized_array
                  size: 4
                  type: f32
          optional: false
    BasicStructWithReadme:
      name: BasicStructWithReadme
      namespace: abc.corp.test
      description: "# test\nthis is a struct description\n"
      version: 1
      fields:
        a:
          name: a
          namespace: abc.corp.test
          description: field a
          version: ~
          type: i32
          optional: false
    SomeStructWithEnum:
      name: SomeStructWithEnum
      namespace: abc.corp.test
      description: ~
      version: 1
      fields:
        enum_value:
          name: enum_value
          namespace: abc.corp.test
          description: ~
          version: ~
          type:
            compound_type:
              kind: enum
              ref: SomeEnum
          optional: false
  enums: {}
